File: RPNCalculator2.hs
-- 逆ポーランド記法電卓 (関数型スタイル2)

module RPNCalculator2 (
  calculate,
  main
) where

import Data.Char (isDigit)
import Data.List (foldl')

-- | 逆ポーランド記法の文字列を読み込み、計算する。
--
-- >>> calculate "1 2 + 3 / 4 - 5 *"
-- -15.0
-- >>> calculate "1 2 + 3 & 4 - 5 *"
-- *** Exception: unsupported operator '&' is used
-- >>> calculate "1 2 + / 3 4 - 5 *"
-- *** Exception: unexpected pattern found
-- >>> calculate ""
-- *** Exception: empty input
calculate :: String -> Double
calculate rpn = case expr of
  [] -> error "empty input"
  _  -> head $ foldl' calc [] expr
  where
    calc stack x
      | all isDigit x = (read x :: Double) : stack
      | otherwise     =
        case stack of
          (y1:y2:ys) -> opeFn x y2 y1 : ys
          _          -> error "unexpected pattern found"
    expr = words rpn

opeFn :: Fractional a => String -> a -> a -> a
opeFn "+" = (+)
opeFn "-" = (-)
opeFn "*" = (*)
opeFn "/" = (/)
opeFn ope = error $ "unsupported operator '" ++ ope ++ "' is used"

-- 利用例
main :: IO ()
main = putStrLn $ rpn ++ "\n = " ++ show (calculate rpn)
  where
    rpn = "1 2 + 3 / 4 - 5 *"

File: rpn_calculator2.erl
%% 逆ポーランド記法電卓 (関数型スタイル2)

-module(rpn_calculator2).
-export([calculate/1]).
-export([main/0]).

calculate(Rpn) ->
  Calc = fun(X, Stack) ->
    case re:run(X, "^[+-]?[0-9]+$") of
      {match, _} ->
        [list_to_integer(X) | Stack];
      _ ->
        case Stack of
          [Y1,Y2|Ys] ->
            OpeFn = ope_fn(X),
            [OpeFn(Y2, Y1) | Ys]
        end
    end
  end,
  Expr = string:tokens(Rpn, " "),
  hd(lists:foldl(Calc, [], Expr)).

ope_fn("+") ->
  fun(X, Y) -> X + Y end;
ope_fn("-") ->
  fun(X, Y) -> X - Y end;
ope_fn("*") ->
  fun(X, Y) -> X * Y end;
ope_fn("/") ->
  fun(X, Y) -> X / Y end;
ope_fn(Ope) ->
  error("unsupported operator '" ++ Ope ++ "' is used").

%% 利用例
main() ->
  Rpn = "1 2 + 3 / 4 - 5 *",
  io:format("~s~n = ~p~n", [Rpn, calculate(Rpn)]).
